[DOCS] Add searchable snapshots topic. (#63040)
elastic/elasticsearch: docs/reference/searchable-snapshots/index.asciidoc

[[searchable-snapshots]]
== {search-snaps-cap}

> 

Reviewer: Add `beta[]` here?

Nodes in a distributed system like {es} will inevitably fail from time to time.
To protect your data against node failures, by default when you index a
document into {es} it is stored on two or more nodes. You also take periodic
<<snapshot-restore,snapshots>> of your data so that you can recover from more
serious failures. This means that each document is stored in at least three
places. These extra copies are important for resiliency, but the storage they
consume has an impact on your cluster's operating costs. The two storage
mechanisms have different, but complementary, performance characteristics:

* Snapshot repositories are much more reliable than local storage on individual
  nodes.

* The monetary cost per GB stored in a snapshot repository is usually lower
  than on a node.

* The monetary cost per read or write operation on a snapshot repository is
  usually much higher.

* Reading or writing data in a snapshot repository usually takes much more time
  compared with accessing a node's local storage.

{search-snaps-cap} let you reduce your operating costs by treating the snapshot
as the authoritative copy of some of your indices. The high reliability of the

> snapshot repository removes the need to keep multiple copies of their data in

Reviewer: `their` seems wrong, did you mean `the`?

your cluster purely for resiliency. {es} makes a copy of a searchable snapshot
on the nodes in the cluster to reduce the performance impact and costs of
accessing the snapshot repository.

With {search-snaps-cap} you may be able to halve your cluster size without
increasing the risk of data loss or reducing the amount of data exposed to
searches. Put differently, {search-snaps-cap} may allow you to expose twice as
much data to searches for a given cluster size.

=== Using searchable snapshots

A searchable snapshot can be searched just like any other index.
{search-snaps-cap} are often used to access a large archive of historical data,

> for which searches may sometimes be complex and time-consuming.

Reviewer: I wonder if we are here framing searchable snapshots as being slow, which is not necessarily the case. I think I get that the time span of the search will be large and this adds to query time, hence async search is necessary, but I would like to move that to the meat of the section more than this introducing paragraph.

Reviewer 2: I was trying not to, but yes it still suggests slowness. I moved this to a `TIP` in 8beb64bc81451f5953bdb8c1a2626857e23ac8e3 and replaced this sentence with one indicating that performance should be similar to a regular index.

<<async-search>> is particularly useful for these long-running searches.

The shards of searchable snapshots are also allocated just like shards of any
other index. You can, for instance, use <<shard-allocation-filtering>> to
restrict these shards to a subset of your nodes.

> Normally you will use {search-snaps-cap} via the

Reviewer: "use" implies "search" in my head, I would prefer to say "create" or "mount" here?

Reviewer 2: I wanted to talk about something more general than just creating them -- after all ILM does also take care of aliases which lets you search them too. How about `manage`? 90a2184f94fb078b0db8858fd593b4829a349266

a large number of indices, most of which will not be mounted as searchable
snapshots. If you mount just one of these indices then you may not delete that
snapshot. Therefore we recommend that you use the <<clone-snapshot-api>> to
cheaply create a clone of a snapshot where the clone contains just the index
you want to mount. This will mean you can delete the multiple-index snapshot,
reducing the size of your snapshot repository, without losing access to any
mounted indices.

We recommend that you <<indices-forcemerge,force-merge>> indices to a single
segment per shard before mounting them as searchable snapshots. Each read from
a snapshot repository takes time and costs money, and the fewer segments there
are the fewer reads are needed to restore the snapshot.

By default a searchable snapshot has `number_of_replicas` set to `0`. You can
increase the number of replicas if desired, for instance if you want to perform
more concurrent searches of these shards.

If a node fails while holding some zero-replica searchable snapshot then there
will be a brief window of time before {es} allocates these shards elsewhere.

> During this window of time the cluster health will be `red` and searches that

Reviewer: I believe the cluster will currently only be yellow unless we worked on this recently?

Reviewer 2: Ah yes now that the recovery source is always "snapshot" this should be the case. Hedging my bets in c43841dc1d59ac7dd814a3332818501619996617 by saying "not green".

converts your index into a {search-snap} when it reaches the `cold` ILM phase.
If you already have some snapshots that you want to search, you can also use
the <<searchable-snapshots-api-mount-snapshot, mount snapshot API>> to manually
mount them as {search-snaps}.

You must not delete a snapshot while any of its indices are mounted as a
{search-snap}. However, most snapshots are taken as a backup and contain a
large number of indices, most of which will not be mounted as searchable
snapshots. If you mount just one of these indices then you may not delete that
snapshot. Therefore we recommend that you use the <<clone-snapshot-api, clone
snapshot API>> to cheaply create a clone of a snapshot where the clone contains
just the index you want to mount. This will mean you can delete the
multiple-index snapshot, reducing the size of your snapshot repository, without
losing access to any mounted indices.

We recommend that you <<indices-forcemerge, force-merge>> indices to a single
segment per shard before mounting them as {search-snaps}. Each read from a
snapshot repository takes time and costs money, and the fewer segments there

> are the fewer reads are needed to restore the snapshot.

Reviewer 2: I'd rather this was down here. It's not _very_ important to force-merge things before mounting them, and if you're mounting an existing snapshot you basically have no choice since you can't do anything about the segment count without restoring each index, merging it and re-snapshotting it.

> By default a {search-snap} index has `number_of_replicas` set to `0`. You can
> increase the number of replicas if desired, for instance if you want to perform
> more concurrent searches of these shards.
> 
> If a node fails while holding some shards of a zero-replica {search-snap} then
> there will be a brief window of time before {es} allocates these shards
> elsewhere.  During this window of time the cluster health will not be `green`
> and searches that hit these shards will fail or return partial results.

Author: Moved up

```suggestion

```

[TIP]
====

> {search-snaps-cap} are ideal for managing a large archive of historical data.
> Historical information is typically searched less frequently than recent data
> and therefore may not need replicas for their performance benefits.
> 
> You can use <<async-search>> with {search-snaps}, which is especially useful
> for more complex or time-consuming searches.

Author: ```suggestion
{search-snaps-cap} are ideal for managing large archives of historical data.
Historical information is typically searched less frequently than recent data
and performance is less important.

For more complex or time-consuming searches, you can use <<async-search>> with {search-snaps}.
```

Reviewer 2: I'd rather keep this wording as-is:

- "performance is less important" suggests to me that there's a general performance penalty for using searchable snapshots -- in fact the only drawback most of the time is the lack of replicas.

I'll apply the change to the wording re. async searches separately.

====

=== How {search-snaps} work

> 
> When you mount a {search-snap} index, {es} allocates its shards onto the data
> nodes in your cluster similarly to shards of regular indices. When a shard of a
> {search-snap} index is allocated to a data node, that node automatically
> restores the shard data from the repository into its local storage. When the
> restore process has completed these shards will respond to searches using the
> data held in local storage and will not need to access the repository. This
> avoids incurring the monetary cost or performance penalty associated with
> reading data from the repository. However, if the node holding one of these
> shards fails then {es} will automatically allocate the shards onto the other
> nodes in the cluster and restore the shard data from the repository again. This
> means you can safely run these indices without replicas, and yet you do not
> need to perform any complicated monitoring or orchestration to restore lost
> shards yourself.

Author: ```suggestion
When an index is mounted from a snapshot, {es} allocates its shards to data
nodes within the cluster. The data nodes then automatically restore the shard data from the repository into local storage. 
Once the restore process completes, these shards respond to searches using the
data held in local storage and do not need to access the repository. This
avoids incurring the cost or performance penalty associated with
reading data from the repository. 

If a node holding one of these shards fails, {es} automatically allocates it to another node,
and that node restores the shard data from the repository. No replicas are needed, and 
no complicated monitoring or orchestration is necessary to  restore lost shards.
```

> Restoring a shard of a {search-snap} index happens in the background, which
> means that you can search these shards even if they have not been fully
> restored. If you attempt to search a shard of a {search-snap} index before it
> has been fully restored then {es} will eagerly retrieve just the data needed
> for the search. This means that some searches will be slower if the shard is
> freshly allocated to a node and still warming up. Searches usually only need to
> access a very small fraction of the total shard data so the performance penalty
> on searches during the background restore process is often very small.

Author: ```suggestion
{es] restores {search-snap} shards in the background and you can search them 
even if they have not been fully restored. If a search hits a {search-snap} shard before it
has been fully restored, {es} eagerly retrieves the data needed for the search. 
If a shard is freshly allocated to a node and still warming up, some searches will be slower. However, searches typically access a very small fraction of the total shard data so the performance penalty is typically very small.
```

[[searchable-snapshots]]
== {search-snaps-cap}

{search-snaps-cap} enable you to significantly reduce costs by 
leveraging external storage for read-only data. 
Searchable snapshots can be used in two ways:

* Snapshot-backed indices: Instead of maintaining replicas of your data within the cluster, 
a snapshot-backed index relies on a searchable snapshot for redundancy. 
You can perform all regular data retrieval operations, but cannot write to the index. 
In the event of a failure, data is recovered from the snapshot. 

> Latency increases during recovery, but you can continue to query your data.

Reviewer: Maybe add that searchable snapshots are eagerly downloaded and once fully downloaded, search performance against searchable snapshots will be similar to search performance against a regular index.

* Fully-remote storage: Data is only brought into the cluster from the snapshot 
when it is needed to service a query. 
Because of the additional overhead, querying your data takes significantly longer 

> and it’s only accessible via asynchronous search.

Reviewer: I believe this option is part of our future plans, but not yet available.

[[searchable-snapshots]]
== {search-snaps-cap}

beta::[]

{search-snaps-cap} let you reduce your operating costs by using
<<snapshot-restore, snapshots>> for resiliency rather than maintaining
<<scalability,replica shards>> within a cluster. When you mount an index from a
snapshot as a {search-snap}, {es} copies the index shards to local storage
within the cluster. This ensures that search performance is comparable to
searching any other index, and minimizes the need to access the snapshot
repository. Should a node fail, shards of a {search-snap} are automatically
recovered from the snapshot repository.

This can result in significant cost savings. With {search-snaps}, you may be
able to halve your cluster size without increasing the risk of data loss or
reducing the amount of data you can search. Because {search-snaps} rely on the
same snapshot mechanism you use for backups, they have a minimal impact on your
snapshot repository storage costs.

[discrete]
[[using-searchable-snapshots]]
=== Using {search-snaps}

> Searching a {search-snap} is the same as searching any other index. Search

Reviewer: Should this be "Searching a searchable snapshot index" to be consistent with glossary. I think it reads better too.

Same comment goes for a number of the "{search-snap}" mentions throughout.

Reviewer 2: Good point, I've added a few "index" or "shard" nouns throughout in 990707b6758d4c82243aec92cb582e34daa828ca.

performance is comparable to regular indices because the shard data is copied
onto nodes in the cluster when the {search-snap} is mounted.

You can control the allocation of the shards of {search-snap} indices using the
same mechanisms as for regular indices. For example, you could use
<<shard-allocation-filtering>> to restrict {search-snap} shards to a subset of

> your nodes.

Reviewer 2: I'm ambivalent -- I put it here since we're starting off by talking about how these indices are mostly maniuplated (searched & allocated) as if they were normal indices, but I've moved it in 990707b6758d4c82243aec92cb582e34daa828ca.

====

[discrete]
[[how-searchable-snapshots-work]]
=== How {search-snaps} work

When an index is mounted from a snapshot, {es} allocates its shards to data
nodes within the cluster. The data nodes then automatically restore the shard
data from the repository onto local storage. Once the restore process
completes, these shards respond to searches using the data held in local
storage and do not need to access the repository. This avoids incurring the
cost or performance penalty associated with reading data from the repository.

If a node holding one of these shards fails, {es} automatically allocates it to
another node, and that node restores the shard data from the repository. No
replicas are needed, and no complicated monitoring or orchestration is
necessary to restore lost shards.

{es} restores {search-snap} shards in the background and you can search them
even if they have not been fully restored. If a search hits a {search-snap}
shard before it has been fully restored, {es} eagerly retrieves the data needed
for the search. If a shard is freshly allocated to a node and still warming up,
some searches will be slower. However, searches typically access a very small

> fraction of the total shard data so the performance penalty is typically very

Reviewer: nit: I think "very" is unnecessary here.

Reviewer 2: Very well ;) Addressed in 990707b6758d4c82243aec92cb582e34daa828ca
